types/trigger.d.ts

```diff
@@ -43,34 +46,110 @@ export type Output = {
-}
-
-export type TimelineTrigger = Trigger & { beforeSeconds: number };
+export type ResponseField = ResponseFunc<MatchesAny> | ResponseOutput<MatchesAny>;
 
 export type TriggerAutoConfig = {
   Output?: Output;
   Duration?: number;
   BeforeSeconds?: number;
   OutputStrings?: { [outputKey: string]: Lang };
 }
+
+export type MatchesAny = { [s in T]?: string } | undefined;
+export type TriggerField<Return> = TriggerFunc<MatchesAny, Return> | Return;
+
+// TODO: I am not sure that it is possible to type triggers such that we have some gigantic union
+// of Trigger<TetherParams> | Trigger<AbilityParams> | Trigger<AddedCombatantParams> | etc etc
+// and so the idea is that individual files would use a narrower type for Matches if they wanted to
+// and that's something we (maybe?) could enforce via tests rather than via TypeScript.  This
+// approach would not preclude some TypeScript gigabrain solution that can figure out how to do
+// the same thing at the TypeScript level.
+//
+// TODO: I think there are two types of TriggerSets.  There is the "cactbot + user defined triggers"
+// and also the "cactbot built-in only" triggers, which can be more strict.  For example,
+// * is `id` required, and `regex` for timeline triggers? (yes for built-in, no for user triggers)
+// * should we handle nonsense like `condition` returning non-booleans?
+// We could have one type that built-in triggers define themselves as more strictly and then a
+// second type that popup-text.js uses and needs to do more narrowing for.
+export type BaseTrigger = {
+  id?: string;
+  disabled?: boolean;
+  condition?: TriggerField<boolean | undefined>;
+  preRun?: TriggerField<void>;
+  delaySeconds?: TriggerField<number | undefined>;
+  durationSeconds?: TriggerField<number | undefined>;
+  suppressSeconds?: TriggerField<number | undefined>;
+  promise?: TriggerField<Promise<void> | undefined>;
+  sound?: TriggerField<string | undefined>;
+  soundVolume?: TriggerField<number | undefined>;
+  response?: ResponseField;
+  alarmText?: TriggerField<TriggerOutput>;
+  alertText?: TriggerField<TriggerOutput>;
+  infoText?: TriggerField<TriggerOutput>;
+  tts?: TriggerField<TriggerOutput>;
+  run?: TriggerField<void>;
+  outputStrings?: {
+    [key: string]: LocaleText;
+  };
+}
+
+export type NetRegexTrigger = BaseTrigger & {
+  netRegex: RegExp;
+  netRegexDe?: RegExp;
+  netRegexFr?: RegExp;
+  netRegexJa?: RegExp;
+  netRegexCn?: RegExp;
+  netRegexKo?: RegExp;
+}
+
+export type RegexTrigger = BaseTrigger & {
+  regex: RegExp;
+  regexDe?: RegExp;
+  regexFr?: RegExp;
+  regexJa?: RegExp;
+  regexCn?: RegExp;
+  regexKo?: RegExp;
+}
+
+export type TimelineTrigger = BaseTrigger & {
+  regex?: RegExp;
+  // TODO: can this also be a function?
+  beforeSeconds?: number;
+};
+
+export type Trigger = RegexTrigger | NetRegexTrigger;
+
+export type TimelineFunc = string | string[] | ((data: Data) => TimelineFunc);
+
+export type TriggerSet = {
+  zoneId?: ZoneId | ZoneId[];
```

Comment:
this becomes `number | null | Array<number | null>`

Reply:
I think that's technically correct? `zoneId` is either a single `ZoneId` or an array of `ZoneIds`.

I am not sure what you are suggesting.  Do you think it should be `number | null | number[]`?

Reply:
Yes, I think it should be `number | null | number[]`. Array with null as an element is unnecessary.

Reply:
Done, thanks!